tests/debug_nans_test.py

```diff
@@ -52,11 +54,29 @@ def testJitComputationNoNaN(self):
     ans.block_until_ready()
 
   def testJitComputationNaN(self):
+    # we warm up the compilation cache for div; otherwise this test depends on
+    # the compilation cache state
```

Comment:
Can you say more? I'm not following why the test depends on the cache warmth (and why it should).

Reply:
Good idea. I'll improve the comment. Here's a first draft of an explanation.

As you know, the situation is different depending on the jaxlib version.

For jaxlib==0.1.64 (current on pypi), nan checks are called for `api.jit` [from the Python wrapper code](https://github.com/google/jax/blob/640e62c7dab22b582a729de7b2eea1ce2c6b480d/jax/api.py#L385-L416), whereas for jaxlib>=0.1.65, [nan checks are called from the C++ dispatch code](https://github.com/google/jax/blob/640e62c7dab22b582a729de7b2eea1ce2c6b480d/jax/api.py#L127-L144).

An analogous situation holds with this op-by-op dispatch stuff as well: for jaxlib==0.1.64, [we need to call nan checks in the Python wrapper](https://github.com/google/jax/pull/6132/files#diff-121975a1a88d61a259f8423655e23a1b389eb18c399b267395d2aba5a196bae9R276), but after 0.1.65 those will be redundant and [we can delete them](https://github.com/google/jax/pull/6132/files#diff-121975a1a88d61a259f8423655e23a1b389eb18c399b267395d2aba5a196bae9R263).

Until we delete them, i.e. until 0.1.65, there are _two_ places which might raise a `FloatingPointException` for a nan error. Which one we hit depends on whether we've already compiled an appropriate version for the primitive that raises an error (e.g. `div` with `f32[]` arguments).

The first place is [in `_execute_compiled`](https://github.com/google/jax/blob/640e62c7dab22b582a729de7b2eea1ce2c6b480d/jax/interpreters/xla.py#L349), called inside [`_cache_miss`](https://github.com/google/jax/pull/6132/files#diff-121975a1a88d61a259f8423655e23a1b389eb18c399b267395d2aba5a196bae9R293). We'll hit that call to `check_special` if we have an empty compilation cache when we compile and run the nan-producing computation (since that first time we dispatch via `_execute_compiled`).

If the compilation cache is populated, though, we'll just jump to the computation from C++, so we'll never hit `_execute_compiled`. Instead, we have to raise the error [from the Python wrapper](https://github.com/google/jax/pull/6132/files#diff-121975a1a88d61a259f8423655e23a1b389eb18c399b267395d2aba5a196bae9R276).

This test and the new one I added just after it together exercise both paths.

At jaxlib==0.1.64, since we have the check in the Python wrapper, we don't actually need the check in `_cache_miss`, i.e. the one in `_execute_compiled`. But to avoid it we'd have to avoid or change the current `_execute_compiled` (i.e. `xla_callable`) standard code path. Moreover, at jaxlib==0.1.65, we don't want the check in the Python wrapper; we want it in the `_cache_miss` path. So I think it's easiest to keep the checks redundant in jaxlib==0.1.64, as it means simpler code now and easy-to-update code later.

Phew, that was complicated. Not sure if I over-complicated it, or if there's just a better way to explain the complexity...

Reply:
I added this comment:

```
# See the subsequent test as well. This test covers the case where we get a
# Python compilation cache hit, since the next line ensures an appropriate
# version of the primitive impl rule is compiled. After jaxlib==0.1.65 both
# tests likely hit the same code paths, but at jaxlib==0.1.64 they cover
# different cases (namely whether we hit the nan check in
# `xla._execute_compiled` or the one in `xla.apply_primitive`).
```

Reply:
We just made jax==0.1.65 the minimum version, so I can clean this up and these long comments will no longer be necessary!